### PR TITLE
Thunder beads

### PR DESCRIPTION
**Pull request by ThatDamnPineapple**
Adds [Thunder beads](https://trello.com/c/KXOUhiXh/376-thunder-beads-2), now called Magnet beads. Also creates a new item directory called Magnet, and moves the thunderbuss there.

Still have to make the whole magnet set obtainable.

### CODE REVIEW

File: Content/Items/Magnet/Weapons.ThunderBeads.cs [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+using StarlightRiver.Helpers;
+using System;
+using System.Collections.Generic;
+using Terraria.Audio;
+using Terraria.GameContent.Creative;
+using Terraria.ID;
+using Terraria.Graphics.Effects;
+using static System.Windows.Forms.VisualStyles.VisualStyleElement.TaskbarClock;
+using StarlightRiver.Content.Dusts;
+using System.Runtime.InteropServices;
```
Comment (ScalarVector1) [inline Content/Items/Magnet/Weapons.ThunderBeads.cs:12]: What is this doing here?

File: Content/Items/Magnet/Weapons.ThunderBeads.cs [hunk 2]
```diff
@@ -0,0 +1,280 @@
+﻿using log4net.Core;
+using ReLogic.Content;
+using StarlightRiver.Helpers;
+using System;
+using System.Collections.Generic;
+using Terraria.Audio;
+using Terraria.GameContent.Creative;
+using Terraria.ID;
+using Terraria.Graphics.Effects;
+using static System.Windows.Forms.VisualStyles.VisualStyleElement.TaskbarClock;
```
Comment (ScalarVector1) [inline Content/Items/Magnet/Weapons.ThunderBeads.cs:10]: What is this doing here?

File: Content/Items/Magnet/Weapons.ThunderBeads.cs [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+namespace StarlightRiver.Content.Items.Magnet
+{
+	public class ThunderBeads : ModItem
+	{
+		public override string Texture => AssetDirectory.MagnetItem + Name;
+
+		public override void SetStaticDefaults()
+		{
+			DisplayName.SetDefault("Magnet Beads");
+			Tooltip.SetDefault("Whip enemies to stick the beads in them \nRepeatedly click to shock affected enemies");
```
Comment (ScalarVector1) [inline Content/Items/Magnet/Weapons.ThunderBeads.cs:23]: We could probably do with some less suggestive wording for this...

File: Content/Items/Magnet/Weapons.ThunderBeads.cs [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+		public override void SetDefaults()
+		{
+			Item.DefaultToWhip(ModContent.ProjectileType<ThunderBeads_Whip>(), 15, 1.2f, 5f, 25);
+			Item.value = Item.sellPrice(0, 2, 0, 0);
+			Item.rare = ItemRarityID.Orange;
+		}
+	}
+
+	public class ThunderBeads_Whip : BaseWhip
```
Comment (ScalarVector1) [inline Content/Items/Magnet/Weapons.ThunderBeads.cs:35]: The underscore here violates naming convention, this should likely be changed to ThunderBeadsProjectile or something similar

File: Content/Items/Magnet/Weapons.ThunderBeads.cs [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+				this.target = target;
+				embedded = true;
+			}
+			else
+			{
+				Helper.PlayPitched("Magic/LightningShortest1", 0.5f, Main.rand.NextFloat(0f, 0.2f), target.Center);
+				fade = 1;
+				for (int i = 2; i < cache.Count - 2; i++)
+				{
+					Dust.NewDustPerfect(cache[i] + Main.rand.NextVector2Circular(6,6), ModContent.DustType<GlowLineFast>(), cache[i].DirectionTo(Owner.Center).RotatedByRandom(0.2f) * Main.rand.NextFloat(1, 2), 0, Color.Cyan, 0.4f);
```
Comment (ScalarVector1) [inline Content/Items/Magnet/Weapons.ThunderBeads.cs:160]: Might be worth adding a linebreak here or pull out a local for the velocity

File: Content/Items/Magnet/Weapons.ThunderBeads.cs [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+			cache = new List<Vector2>();
+			SetPoints(cache);
+
+			cache2 = new List<Vector2>();
+			for (int i = 0; i < cache.Count; i++)
+			{
+				Vector2 point = cache[i];
+				Vector2 endPoint = embedded ? target.Center : cache[i];
+				Vector2 nextPoint = i == cache.Count - 1 ? endPoint : cache[i + 1];
+				Vector2 dir = Vector2.Normalize(nextPoint - point).RotatedBy(Main.rand.NextBool() ? -1.57f : 1.57f);
```
Comment (ScalarVector1) [inline Content/Items/Magnet/Weapons.ThunderBeads.cs:221]: Linebreak here would help clarity

File: Content/Items/Magnet/Weapons.ThunderBeads.cs [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+			trail ??= new Trail(Main.instance.GraphicsDevice, segments + 1, new TriangularTip(4), factor => 16, factor =>
+			{
+				if (factor.X > 0.99f)
+					return Color.Transparent;
+
+				return new Color(160, 220, 255) * fade * 0.1f * EaseFunction.EaseCubicOut.Ease(1 - factor.X);
+			});
+
+			trail.Positions = cache.ToArray();
+			trail.NextPosition = endPoint;
```
Comment (ScalarVector1) [inline Content/Items/Magnet/Weapons.ThunderBeads.cs:241]: Might be worth it to put a linebreak between the areas working on trail and trail2 for clarity

File: Core/BaseWhip.cs [hunk 8]
```diff
@@ -105,7 +108,7 @@ public override void CutTiles()
 			}
 		}
 
-		public void SetPoints(List<Vector2> controlPoints)
+		public virtual void SetPoints(List<Vector2> controlPoints)
```
Comment (ScalarVector1) [inline Core/BaseWhip.cs:111]: Good change

**Review by ScalarVector1 — CHANGES_REQUESTED** [review]

File: Content/Items/Magnet/Weapons.ThunderBeads.cs (hunk 1 above)
Comment (ThatDamnPineapple) [reply Content/Items/Magnet/Weapons.ThunderBeads.cs:12]: Fixed

File: Content/Items/Magnet/Weapons.ThunderBeads.cs (hunk 2 above)
Comment (ThatDamnPineapple) [reply Content/Items/Magnet/Weapons.ThunderBeads.cs:10]: Fixed

File: Content/Items/Magnet/Weapons.ThunderBeads.cs (hunk 3 above)
Comment (ThatDamnPineapple) [reply Content/Items/Magnet/Weapons.ThunderBeads.cs:23]: Fixed

File: Content/Items/Magnet/Weapons.ThunderBeads.cs (hunk 4 above)
Comment (ThatDamnPineapple) [reply Content/Items/Magnet/Weapons.ThunderBeads.cs:35]: Fixed

File: Content/Items/Magnet/Weapons.ThunderBeads.cs (hunk 5 above)
Comment (ThatDamnPineapple) [reply Content/Items/Magnet/Weapons.ThunderBeads.cs:160]: Fixed

File: Content/Items/Magnet/Weapons.ThunderBeads.cs (hunk 6 above)
Comment (ThatDamnPineapple) [reply Content/Items/Magnet/Weapons.ThunderBeads.cs:221]: Fixed

File: Content/Items/Magnet/Weapons.ThunderBeads.cs (hunk 7 above)
Comment (ThatDamnPineapple) [reply Content/Items/Magnet/Weapons.ThunderBeads.cs:241]: Fixed

**Review by ScalarVector1 — APPROVED** [review]

**Pull request merged**